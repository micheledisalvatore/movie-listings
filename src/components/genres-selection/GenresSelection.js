import React from 'react';
import PropTypes from 'prop-types';
import { values } from 'ramda';
import { bindActionCreators } from 'redux';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';

import { setFilterGenres } from '../../actions/genres';

const ITEM_HEIGHT = 48;
const ITEM_PADDING_TOP = 8;
const MenuProps = {
  PaperProps: {
    style: {
      maxHeight: ITEM_HEIGHT * 4.5 + ITEM_PADDING_TOP,
      width: 250,
    },
  },
};

export const GenresSelection  = ({
  setFilterGenresAction,
  classes,
  genres,
  filters,
}) => {
  const handleChange = event => {
    const { target: { value } } = event;

    setFilterGenresAction(value);
  };

  return (
    <div className={classes.root}>
      <InputLabel htmlFor="select-multiple" className={classes.label}>Genres:</InputLabel>
      <FormControl className={classes.formControl}>
        <Select
          multiple
          value={filters}
          onChange={handleChange}
          input={<Input id="select-multiple" />}
          MenuProps={MenuProps}
        >
          <MenuItem disabled value="">
            <em>All</em>
          </MenuItem>
          {genres.map(({ id, name }) => (
            <MenuItem key={id} value={id}>
              {name}
            </MenuItem>
          ))}
        </Select>
      </FormControl>
    </div>
  );
}

GenresSelection.propTypes = {
  classes: PropTypes.object.isRequired,
  genres: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      name: PropTypes.string,
    }),
  ).isRequired,
  filters: PropTypes.arrayOf(PropTypes.number).isRequired,
  setFilterGenresAction: PropTypes.func.isRequired,
};

export const mapStateToProps = ({ genres: { list, filters }}) => ({
  genres: values(list).sort((a, b) => a.name.localeCompare(b.name)) || [],
  filters,
});

export const mapDispatchToProps = dispatch => ({
  setFilterGenresAction: bindActionCreators(setFilterGenres, dispatch),
});
