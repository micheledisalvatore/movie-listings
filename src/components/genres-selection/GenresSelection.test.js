/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';

import { GenresSelection, mapStateToProps } from './GenresSelection';

describe('Given a GenresSelection component', () => {
  let component;
  let wrapper;
  const setFilterGenresActionMock = jest.fn();
  const requiredProps = {
    classes: {},
    genres: [],
    filters: [],
    setFilterGenresAction: setFilterGenresActionMock,
  };

  describe('when it is rendered', () => {
    beforeEach(() => {
      component = <GenresSelection {...requiredProps} />;
      wrapper = shallow(component);
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the placeholder', () => {
      expect(wrapper.find(MenuItem).at(0).find('em')).toHaveText('All');
    });

    it('should NOT render other options', () => {
      expect(wrapper.find(MenuItem)).toHaveLength(1);
    })

    describe('and a list of genres is passed', () => {
      beforeEach(() => {
        wrapper.setProps({
          genres: [{
            id: 0,
            name: 'foo',
          },{
            id: 1,
            name: 'bar',
          }]
        });
      });

      it('should match the snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render the passed options', () => {
        expect(wrapper.find(MenuItem)).toHaveLength(3);
      });

      it('should render the second option with the passed id', () => {
        expect(wrapper.find(MenuItem).at(2)).toHaveProp('value', 1);
      });

      describe('and the user selects the first option', () => {
        beforeEach(() => {
          wrapper.find(Select).simulate('change', { target: { value: [0] }});
        });

        it('should call setFilterGenresAction with the selected options', () => {
          expect(setFilterGenresActionMock).toHaveBeenCalledWith([0]);
        });

        describe('and the list of filters is updated', () => {
          beforeEach(() => {
            wrapper.setProps({
              filters: [0],
            });
          });

          it('should render the Select component with the passed filters', () => {
            expect(wrapper.find(Select)).toHaveProp('value', [0]);
          });
        })
      })
    })
  });
});

describe('Given a mapStateToProps function', () => {
  describe('when it is called', () => {
    it('should return an object with the passed values', () => {
      expect(mapStateToProps({
        genres: {
          list: {
            1: {
              id: 1,
              name: 'foo',
            },
            3: {
              id: 3,
              name: 'bar',
            },
          },
          filters: ['foo'],
        },
      })).toEqual({
        genres: [{
            id: 3,
            name: 'bar',
          }, {
            id: 1,
            name: 'foo',
          }
        ],
        filters: ['foo'],
      });
    });
  });
});
