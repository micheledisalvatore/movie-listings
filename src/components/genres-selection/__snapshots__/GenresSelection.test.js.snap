// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a GenresSelection component when it is rendered and a list of genres is passed should match the snapshot 1`] = `
<div>
  <WithStyles(WithFormControlContext(InputLabel))
    htmlFor="select-multiple"
  >
    Genres:
  </WithStyles(WithFormControlContext(InputLabel))>
  <WithStyles(FormControl)>
    <WithStyles(WithFormControlContext(Select))
      MenuProps={
        Object {
          "PaperProps": Object {
            "style": Object {
              "maxHeight": 224,
              "width": 250,
            },
          },
        }
      }
      input={
        <WithStyles(Input)
          id="select-multiple"
        />
      }
      multiple={true}
      onChange={[Function]}
      value={Array []}
    >
      <WithStyles(MenuItem)
        disabled={true}
        value=""
      >
        <em>
          All
        </em>
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="0"
        value={0}
      >
        foo
      </WithStyles(MenuItem)>
      <WithStyles(MenuItem)
        key="1"
        value={1}
      >
        bar
      </WithStyles(MenuItem)>
    </WithStyles(WithFormControlContext(Select))>
  </WithStyles(FormControl)>
</div>
`;

exports[`Given a GenresSelection component when it is rendered should match the snapshot 1`] = `
<div>
  <WithStyles(WithFormControlContext(InputLabel))
    htmlFor="select-multiple"
  >
    Genres:
  </WithStyles(WithFormControlContext(InputLabel))>
  <WithStyles(FormControl)>
    <WithStyles(WithFormControlContext(Select))
      MenuProps={
        Object {
          "PaperProps": Object {
            "style": Object {
              "maxHeight": 224,
              "width": 250,
            },
          },
        }
      }
      input={
        <WithStyles(Input)
          id="select-multiple"
        />
      }
      multiple={true}
      onChange={[Function]}
      value={Array []}
    >
      <WithStyles(MenuItem)
        disabled={true}
        value=""
      >
        <em>
          All
        </em>
      </WithStyles(MenuItem)>
    </WithStyles(WithFormControlContext(Select))>
  </WithStyles(FormControl)>
</div>
`;
